packages/plugin-ext-headless/src/hosted/node/plugin-host-headless-rpc.ts

```diff
@@ -0,0 +1,94 @@
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { dynamicRequire } from '@theia/core/lib/node/dynamic-require';
+import { ContainerModule, injectable, inject } from '@theia/core/shared/inversify';
+import { EnvExtImpl } from '@theia/plugin-ext/lib/plugin/env';
+import { LocalizationExt } from '@theia/plugin-ext';
+import { LocalizationExtImpl } from '@theia/plugin-ext/lib/plugin/localization-ext';
+import { HEADLESSMAIN_RPC_CONTEXT } from '../../common/headless-plugin-rpc';
+import { HeadlessPluginManagerExtImpl } from '../../plugin/headless-plugin-manager';
+import { AbstractPluginHostRPC, ExtInterfaces } from '@theia/plugin-ext/lib/hosted/node/plugin-host-rpc';
+import { PluginHost } from '@theia/plugin-ext/lib/plugin/plugin-manager';
+import { PluginModel } from '@theia/plugin-ext/lib/common/plugin-protocol';
+import { ExtPluginApi, ExtPluginApiHeadlessInitializationFn } from '../../common/plugin-ext-headless-api-contribution';
+import { RPCProtocol } from '@theia/plugin-ext/lib/common/rpc-protocol';
+
+type HeadlessExtInterfaces = Pick<ExtInterfaces, 'envExt'|'localizationExt'>;
+
+/**
+ * The RPC handler for headless plugins.
+ */
+@injectable()
+export class HeadlessPluginHostRPC extends AbstractPluginHostRPC<HeadlessPluginManagerExtImpl, undefined, HeadlessExtInterfaces> {
+    @inject(EnvExtImpl)
+    protected readonly envExt: EnvExtImpl;
+
+    @inject(LocalizationExt)
+    protected readonly localizationExt: LocalizationExtImpl;
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    constructor(@inject(RPCProtocol) _rpc: RPCProtocol) {
+        super(
+            _rpc, 'HEADLESSPLUGIN_HOST', undefined,
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            (pluginHost: PluginHost, extInterfaces: HeadlessExtInterfaces, rpc: any) => {
+                const { envExt, localizationExt } = extInterfaces;
+                return new HeadlessPluginManagerExtImpl(pluginHost,
+                    envExt, localizationExt,
+                    rpc
+                );
+            },
+            {
+                $pluginManager: HEADLESSMAIN_RPC_CONTEXT.HOSTED_PLUGIN_MANAGER_EXT,
+            }
+        );
+    }
+
+    protected createExtInterfaces(): HeadlessExtInterfaces {
+        return {
+            envExt: this.envExt,
+            localizationExt: this.localizationExt
+        };
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    protected createAPIFactory(_extInterfaces: HeadlessExtInterfaces): undefined {
+        // As yet there is no default API namespace for backend plugins to access the Theia framework
+        return undefined;
```

Comment:
I wonder if there is anything better that we can return instead of undefined? I understand that the generic argument is also undefined but it feels a but odd, maybe wrongly so.

Reply:
I think perhaps I don't understand the question. We have no API object that we can provide by default, like the `theia` or `vscode` object. I think `null` would better signal that there is not supposed to be an API object rather than it just not yet being available and initialized. Is that what you mean?